### PR TITLE
added an onExit option for tasks such as copying compiled code to an …

### PR DESCRIPTION
**Pull request by liyutech**
…external directory.

First of all, this is a great plugin which we like a lot. This pull request aims to address a corner case described below:

When the onBuildEnd option is run, the directory which contains compiled code may still be empty. This makes plugin tasks such as copying the compiled code to an external directory or ssh it to a remote server more difficult.

When the "done" event is emitted, the compile code directory is populated. That's helpful for our user cases.

### CODE REVIEW

**Comment by opiepj** [comment]
Curious, would this also be a solution to webpack-dev-server's HMR? For instance if you have a script that opens a page, you wouldn't want that script to keep executing for every code change.

File: index.js [hunk 1]
```diff
@@ -38,6 +41,13 @@ WebpackShellPlugin.prototype.apply = function(compiler) {
   }
   callback();
 });
+
+  compiler.plugin("done", function () {
```
Comment (opiepj) [inline index.js]: Does webpack's plugin have an event "done"?

Comment (liyutech) [inline index.js]: yes, according to https://github.com/webpack/docs/wiki/plugins. Here is a relevant part of the doc from that link:


done(stats: Stats)

All is done.

In addition, the new code that use the 'done' event works in our environments.


[https://avatars1.githubusercontent.com/u/2105791?v=3&s=400]<https://github.com/webpack/docs/wiki/plugins>

plugins · webpack/docs Wiki · GitHub<https://github.com/webpack/docs/wiki/plugins>
github.com
For a high-level introduction to writing plugins, start with How to write a plugin. Many objects in Webpack extend the Tapable class, which exposes a plugin method.



[https://avatars1.githubusercontent.com/u/2105791?v=3&s=400]<https://github.com/webpack/docs/wiki/plugins>

plugins · webpack/docs Wiki · GitHub<https://github.com/webpack/docs/wiki/plugins>
github.com
For a high-level introduction to writing plugins, start with How to write a plugin. Many objects in Webpack extend the Tapable class, which exposes a plugin method.





________________________________

**Comment by liyutech** [comment]
Hi, Patrick,


that's an interesting idea which shall work.


Cheers,


Hao


________________________________

File: index.js (hunk 1 above)
Comment (liyutech) [inline index.js]: Here is another example:


HelloWorldPlugin.prototype.apply = function(compiler) {
  compiler.plugin('done', function() {
    console.log('Hello World!');
  });
};




The example is from https://github.com/webpack/docs/wiki/how-to-write-a-plugin

[https://avatars1.githubusercontent.com/u/2105791?v=3&s=400]<https://github.com/webpack/docs/wiki/how-to-write-a-plugin>

how to write a plugin · webpack/docs Wiki · GitHub<https://github.com/webpack/docs/wiki/how-to-write-a-plugin>
github.com
Plugins expose the full potential of the Webpack engine to third-party developers. Using staged build callbacks, developers can introduce their own behaviors into the ...






________________________________

**Comment by opiepj** [comment]
I'll take a look at this tonight and see if I can merge it in, thanks!

**Comment by opiepj** [comment]
Merging your changes #5

**Comment by liyutech** [comment]
cool. Looking forward to cooperating with you later whenever possible.


________________________________

**Pull request closed without merging**